File: cortex/webgraph/openers.py
from __future__ import annotations

import os
import subprocess
import sys
from pathlib import Path


def resolve_safe_vault_path(vault_root: Path, relative_path: str) -> Path:
    candidate = (vault_root / relative_path).resolve()
    root = vault_root.resolve()
    if candidate != root and root not in candidate.parents:
        raise ValueError(f"Refusing to open path outside vault: {candidate}")
    if not candidate.exists():
        raise FileNotFoundError(candidate)
    return candidate


def open_path(path: Path) -> None:
    if sys.platform == "win32":
        os.startfile(str(path))  # type: ignore[attr-defined]
        return
    if sys.platform == "darwin":
        subprocess.run(["open", str(path)], check=False)
        return
    subprocess.run(["xdg-open", str(path)], check=False)

File: rust/crates/cortex-webgraph-server/src/openers.rs
//! Porteo de `cortex/webgraph/openers.py`.

#![forbid(unsafe_code)]

use std::path::{Path, PathBuf};
use std::process::Command;

use cortex_workspace::layout::resolve_lexical;

/// resolve_safe_vault_path: rechaza paths fuera del vault y archivos
/// inexistentes. Errores como Result (Python lanza ValueError/FileNotFoundError).
pub fn resolve_safe_vault_path(vault_root: &Path, relative_path: &str) -> Result<PathBuf, String> {
    let candidate = resolve_lexical(&vault_root.join(relative_path));
    let root = resolve_lexical(vault_root);
    if candidate != root && !candidate.starts_with(&root) {
        return Err(format!(
            "Refusing to open path outside vault: {}",
            candidate.display()
        ));
    }
    if !candidate.exists() {
        return Err(format!("{}", candidate.display()));
    }
    Ok(candidate)
}

/// open_path: xdg-open (Linux) / open (macOS); Windows no soportado acá.
pub fn open_path(path: &Path) {
    let target = path.to_string_lossy().to_string();
    #[cfg(target_os = "macos")]
    let _ = Command::new("open").arg(target).status();
    #[cfg(not(target_os = "macos"))]
    let _ = Command::new("xdg-open").arg(target).status();
}
